### ISSUE TITLE
Failed Canary Analysis run did not stop rollout

### ISSUE THREAD

**Issue by DanTulovsky** [argoproj/argo-rollouts#1411]
## Summary 

Today we managed to push out a bad binary that failed its analysis run, but the rollout did not stop.

As per the logs below, 

```
"time=""2021-08-11T17:59:40Z"" level=info msg=""metric assessed Failed: failed (3) > failureLimit (2)"" metric=router-error-rate-nsfSXCCn
```

which should have aborted the rollout. In the Web UI we saw a completed rollout, but the Analysis bar was red.  This was an automated rollout and nobody did a promote.

```
"time=""2021-08-11T17:59:46Z"" level=info msg=""metric transitioned from Running -> Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:59:47Z"" level=warning msg=""metric 'router-error-rate-nsfSXCCn' completed Failed"" analysisrun=public-router-754f86fd5-18 event_reason=MetricFailed namespace=public
```

You can see the failure message, and then later, you see a success message for the same metric (router-error-rate-nsfSXCCn).

I have a theory. This is what the graph looked like. The red line is error and the black line is ops/second. 

<img width="423" alt="Screen Shot 2021-08-11 at 4 08 24 PM" src="https://user-images.githubusercontent.com/509887/129096214-a6c3bb68-12bc-434d-8d26-9a352b9f4744.png">

(timestamps on the graph are EST)

The failed analysis happened at the large red spike.  Then things broke sufficiently that there was no more traffic (black line is at 0) and so the red error line goes to 0 also.  At which point the analysis succeeds.  And the rollout isn't rolled back.

Might this have something to do with `running overdue measurement` log messages? Is it possible that after the failure, the queued up analysis checks continued to run and, because they started to succeed, everything was reset to "working" again, whereas it should have failed and rolled back?

Thank you

## Diagnostics

1.0.2

# Paste the logs from the rollout controller

```
"time=""2021-08-11T18:57:24Z"" level=info msg=""Event(v1.ObjectReference{Kind:\""AnalysisRun\"", Namespace:\""public\"", Name:\""public-router-74d75cd9bd-19\"", UID:\""df27cfbc-bac3-4fbc-a280-1d4a0af965ae\"", APIVersion         :\""argoproj.io/v1alpha1\"", ResourceVersion:\""3287279442\"", FieldPath:\""\""}): type: 'Normal' reason: 'MetricSuccessful' metric 'router-error-rate-nsfSXCCn' completed Successful""
"time=""2021-08-11T18:57:24Z"" level=info msg=""metric 'router-error-rate-nsfSXCCn' completed Successful"" analysisrun=public-router-74d75cd9bd-19 event_reason=MetricSuccessful namespace=public
"time=""2021-08-11T18:57:23Z"" level=info msg=""metric transitioned from Running -> Successful"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:57:20Z"" level=info msg=""metric assessed Successful: run terminated"" metric=router-error-rate-nsfSXCCn
"time=""2021-08-11T18:57:00Z"" level=info msg=""skipping measurement: run is terminating"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:56:46Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:56:40Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:56:18Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:56:17Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:55:49Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:55:45Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:55:21Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T18:55:11Z"" level=info msg=""running initial measurement"" analysisrun=public-router-74d75cd9bd-19 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:59:47Z"" level=info msg=""Event(v1.ObjectReference{Kind:\""AnalysisRun\"", Namespace:\""public\"", Name:\""public-router-754f86fd5-18\"", UID:\""b47a4433-af35-4b84-8aa1-f1c96c5f30a6\"", APIVersion:\         ""argoproj.io/v1alpha1\"", ResourceVersion:\""3287091312\"", FieldPath:\""\""}): type: 'Warning' reason: 'MetricFailed' metric 'router-error-rate-nsfSXCCn' completed Failed""
"time=""2021-08-11T17:59:47Z"" level=warning msg=""metric 'router-error-rate-nsfSXCCn' completed Failed"" analysisrun=public-router-754f86fd5-18 event_reason=MetricFailed namespace=public
"time=""2021-08-11T17:59:46Z"" level=info msg=""metric transitioned from Running -> Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:59:40Z"" level=info msg=""metric assessed Failed: failed (3) > failureLimit (2)"" metric=router-error-rate-nsfSXCCn
"time=""2021-08-11T17:59:34Z"" level=info msg=""measurement completed Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:59:19Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:58:49Z"" level=info msg=""measurement completed Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:58:33Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:57:46Z"" level=info msg=""measurement completed Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:57:26Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:57:05Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:56:57Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:56:33Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:56:24Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:56:05Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:56:04Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:55:35Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:55:25Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:54:49Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:54:44Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:54:18Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:54:15Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:53:54Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:53:46Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:53:08Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:52:52Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:52:14Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:52:10Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:51:47Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:51:45Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:51:15Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:51:09Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:50:42Z"" level=info msg=""measurement completed Successful"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by alexmt** [comment]
Thank you @DanTulovsky ,

Can you also add rollout spec to simplify reproducing the issue?

**Comment by DanTulovsky** [comment]
```
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  annotations:
    meta.helm.sh/release-name: public-router
    meta.helm.sh/release-namespace: public
    rollout.argoproj.io/revision: "19"
  creationTimestamp: "2021-06-23T13:50:45Z"
  generation: 211
  labels:
    app.kubernetes.io/instance: public
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: router
    app.kubernetes.io/version: 2021-08-04_04-11-05Z
    helm.sh/release: public-router
  name: public-router
  namespace: public
  resourceVersion: "3287586765"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/public/rollouts/public-router
  uid: a646cdd5-c5ec-4bbf-8747-50b15fb588fb
spec:
  progressDeadlineSeconds: 3600
  replicas: 110
  strategy:
    canary:
      analysis:
        startingStep: 1
        templates:
        - templateName: router-errors
      canaryMetadata:
        labels:
          role: canary
      maxSurge: 50%
      maxUnavailable: 0
      stableMetadata:
        labels:
          role: stable
      steps:
      - setWeight: 1
      - pause:
          duration: 600s
      - setWeight: 10
      - pause:
          duration: 600s
      - setWeight: 30
      - pause:
          duration: 600s
      - setWeight: 60
      - pause:
          duration: 900s
  workloadRef:
    apiVersion: apps/v1
    kind: Deployment
    name: public-router
```

**Comment by DanTulovsky** [comment]
Also, there are 6 metrics attached to the AnalysisTemplate all of the form like this:

```
  - failureCondition: result == 'true'
    failureLimit: 2
    interval: 15s
    name: router-error-rate-<id>
    provider:
      web:
        headers:
        - key: Authorization
          value: Bearer {{ args.access-token }}
        jsonPath: '{$.data.attributes.state}'
        timeoutSeconds: 5
        url: <url>
```

**Comment by harikrongali** [comment]
@DanTulovsky with the given config, I am unable to reproduce with the given config. Can you provide me kubectl events that happened during the time frame?

**Comment by DanTulovsky** [comment]
I am afraid those are gone already.  I realize this may be difficult to reproduce, but there must be some path in the code where an analysis fails, but the rollout isn't rolled back.  Perhaps that's where it's best to start looking?  I mean, the "rollback release after analysis fails" is core functionality here. :/

**Comment by jessesuen** [comment]
```
"time=""2021-08-11T17:59:40Z"" level=info msg=""metric assessed Failed: failed (3) > failureLimit (2)"" metric=router-error-rate-nsfSXCCn
"time=""2021-08-11T17:59:34Z"" level=info msg=""measurement completed Failed"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
"time=""2021-08-11T17:59:19Z"" level=info msg=""running overdue measurement"" analysisrun=public-router-754f86fd5-18 metric=router-error-rate-nsfSXCCn namespace=public
```

This log/error sequence is definitely suspicious. If the metric was assessed failed, we have no reason to take further measurements and the entire Run should be deemed Failed. Will have to figure out what went wrong through code inspection.

**Comment by DanTulovsky** [comment]
Hi, is there any update on this?  thank you!

**Comment by harikrongali** [comment]
@DanTulovsky fix is released in 1.1.1, can you validate & let us know if it is working for you?

**Comment by jessesuen** [comment]
Let's close this as fixed and reopen as necessary
